### PR TITLE
Support obs_mask in sample statement.

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR ports @fritzo PR for `obs_mask` https://github.com/pyro-ppl/pyro/pull/2772 in Pyro to NumPyro.

### CODE REVIEW

File: numpyro/primitives.py [hunk 1]
```diff
@@ -80,7 +80,31 @@ def __call__(self, *args, **kwargs):
...
+    with numpyro.handlers.mask(mask=(not obs_mask if isinstance(obs_mask, bool) else ~obs_mask)):
+        unobserved = sample(f"{name}_unobserved", fn, **kwargs)
+
+    # Interleave observed and unobserved events.
+    shape = jnp.shape(obs_mask) + (1,) * fn.event_dim
+    batch_mask = jnp.reshape(obs_mask, shape)
+    try:
+        value = jnp.where(batch_mask, observed, unobserved)
+    except RuntimeError as e:
+        if "must match the size of tensor" in str(e):
```
Comment (fritzo) [inline numpyro/primitives.py:96]: Could you port Pyro's [death test](https://github.com/pyro-ppl/pyro/pull/2772/files#diff-84b43e1d1f62fa4c1d721f194efd2ffe0dcea3e07522cc505109a401af596cd9R2266) to NumPyro? I suspect this error string is torch-specific, and you'll need a different error string and possibly exception type for JAX.

Alternatively, you could just omit this fancy error handling for now.

File: numpyro/primitives.py [hunk 2]
```diff
@@ -101,12 +125,20 @@ def sample(name, fn, obs=None, rng_key=None, sample_shape=(), infer=None):
...
+        imputed by sampling. This introduces a latent sample site named ``name
+        + "_unobserved"`` which should be used by guides.
     :return: sample from the stochastic `fn`.
     """
     # if there are no active Messengers, we just draw a sample and return it as expected:
     if not _PYRO_STACK:
         return fn(rng_key=rng_key, sample_shape=sample_shape)
 
+    if obs_mask is not None:
+        return _masked_observe(name, fn, obs, obs_mask, rng_key=None, sample_shape=(), infer=None)
```
Comment (fritzo) [inline numpyro/primitives.py:140]: Shouldn't this pass through arguments? I'm surprised the `rng_key` bit didn't error 😕 
```py
return _masked_observe(name, fn, obs, obs_mask, rng_key=rng_key, sample_shape=sample_shape, infer=infer)
```

**Review by fritzo — CHANGES_REQUESTED** [review]
Thanks for porting this @fehiepsi !

File: numpyro/primitives.py (hunk 1 above)
Comment (fehiepsi) [reply numpyro/primitives.py:96]: Thanks, I'll omit this for simplicity. :)

File: numpyro/primitives.py [hunk 3]
```diff
@@ -80,7 +80,21 @@ def __call__(self, *args, **kwargs):
             return self.fn(*args, **kwargs)
 
 
-def sample(name, fn, obs=None, rng_key=None, sample_shape=(), infer=None):
+def _masked_observe(name, fn, obs, obs_mask, **kwargs):
+    # Split into two auxiliary sample sites.
+    with numpyro.handlers.mask(mask=obs_mask):
+        observed = sample(f"{name}_observed", fn, **kwargs, obs=obs)
+    with numpyro.handlers.mask(mask=(not obs_mask if isinstance(obs_mask, bool) else ~obs_mask)):
```
Comment (fritzo) [inline numpyro/primitives.py:87]: I had this issue too, and ended up using `___ ^ False` for type safe negation

Comment (fehiepsi) [reply numpyro/primitives.py:87]: Nice! Thank you!

File: numpyro/primitives.py (hunk 2 above)
Comment (fehiepsi) [reply numpyro/primitives.py:130]: Sorry, I missed this comment. I'll address this too.

**Review by fritzo — APPROVED** [review]
LGTM

**Comment by fehiepsi** [comment]
Thanks for reviewing! :)

**Pull request merged**